### PR TITLE
[AIRFLOW-2780] Adds IMAP Hook to interact with a mail server

### PR DESCRIPTION
**Pull request by feluelle**
### JIRA
- [x] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW-2780) issues and references them in the PR title.

### Description
- [x] My PR adds the feature to be able to interact with a mail server via the IMAP Protocol. It exposes the following methods:
- `has_mail_attachment`: Can be used in a `Sensor` to poke the mail server for attachments with the given name and returns True if at least one has been found and False if not.
- `retrieve_mail_attachments`: Can be used in an `Operator` to do sth. with the attachments returned as list of tuple (filename, payload).
- `download_mail_attachments`: Can be used in an `Operator` to download the attachment.

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
- `test_connect_and_disconnect`: this tests calls the ImapHook as ContextManager and so it tests for both (on enter/connect and on exit/disconnect)
- `test_has_mail_attachments_found`
- `test_has_mail_attachments_not_found`
- `test_has_mail_attachments_with_regex_found`
- `test_has_mail_attachments_with_regex_not_found`
- `test_retrieve_mail_attachments_found`
- `test_retrieve_mail_attachments_not_found`
- `test_retrieve_mail_attachments_with_regex_found`
- `test_retrieve_mail_attachments_with_regex_not_found`
- `test_retrieve_mail_attachments_latest_only`
- `test_download_mail_attachments_found`
- `test_download_mail_attachments_not_found`
- `test_download_mail_attachments_with_regex_found`
- `test_download_mail_attachments_with_regex_not_found`
- `test_download_mail_attachments_with_latest_only`

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### Documentation
- [x] In case of new functionality, my PR adds documentation that describes how to use it.
    - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

### Code Quality
- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

**Comment by feluelle** [comment]
@ashb would like to have a code review from you or someone else of the maintainers. :)

File: airflow/contrib/hooks/imap_hook.py
Comment (ashb) [inline airflow/contrib/hooks/imap_hook.py:43]: Does this also disconnect? If not it probably should.

File: airflow/contrib/hooks/imap_hook.py
Comment (ashb) [inline airflow/contrib/hooks/imap_hook.py:154]: What it the attachment is called "../../../../../../etc/passwd"? (Or something else that the airflow user actually has permission to write to.

File: airflow/contrib/hooks/imap_hook.py
Comment (ashb) [inline airflow/contrib/hooks/imap_hook.py:115]: Maybe return True/False or number of files downloaded?

**Review by ashb — CHANGES_REQUESTED** [review]

File: airflow/contrib/hooks/imap_hook.py
Comment (feluelle) [reply airflow/contrib/hooks/imap_hook.py:43]: According to [the official docs](https://docs.python.org/2/library/imaplib.html#imaplib.IMAP4.logout) it shutdowns the connection to the server.

File: airflow/contrib/hooks/imap_hook.py
Comment (feluelle) [inline airflow/contrib/hooks/imap_hook.py:158]: btw.. there is an [iter_attachments](https://docs.python.org/3/library/email.message.html#email.message.EmailMessage.iter_attachments) method but this apparently does not work correctly so I made a Mail class that iterates through all assets and checks if it is really a valid attachment.
I hope I can change this in the future.

File: airflow/contrib/hooks/imap_hook.py
Comment (feluelle) [reply airflow/contrib/hooks/imap_hook.py:154]: What do you mean? ..whether it would override a file or not?

File: airflow/contrib/hooks/imap_hook.py
Comment (feluelle) [reply airflow/contrib/hooks/imap_hook.py:115]: Hm.. I dont know. I personally would not return anything because it changes the filesystem. A method should either change something or return something but not both?!

Of course it should raise an exception if there is a problem when downloading and/or writing a file.

File: airflow/contrib/hooks/imap_hook.py
Comment (ashb) [reply airflow/contrib/hooks/imap_hook.py:154]: We should put a check in to ensure that the file we create doesn't "escape" the local_output_directory

Also we should consider what happens if the attachment is a symlink (not sure that's possible in mail but...?)
Not this exact case, but this describes a similar problem that could happen https://rt.cpan.org/Public/Bug/Display.html?id=125523

Comment (feluelle) [reply airflow/contrib/hooks/imap_hook.py:154]: Agree. It should check for "escaping" and check for symlinks.
I will implement this.

**Comment by SamWildmo** [comment]
Will this hook be in 1.10.1?

**Comment by feluelle** [comment]
> Will this hook be in 1.10.1?

I hope so. I will try to get it ready for review tomorrow, containing checking for escaping chars and symlinks.

**Comment by feluelle** [comment]
I am going to close this PR, because I want to open a new one directly from my feature branch instead of my master branch. @ashb @SamWildmo

**Pull request closed without merging**
